src/common/src/com/microsoftopentechnologies/tooling/msservices/helpers/auth/AADManagerImpl.java

```diff
@@ -0,0 +1,697 @@
+
+        ReentrantReadWriteLock userLock = getUserLock(userInfo, false);
+        userLock.writeLock().lock();
+
+        try {
+            if (AuthenticationResult.equals(refreshAuthResult, getRefreshAuthResult(userInfo))) {
+                PluginSettings settings = DefaultLoader.getPluginComponent().getSettings();
+                AuthenticationContext context = null;
+
+                try {
+                    context = new AuthenticationContext(settings.getAdAuthority());
+                    AuthenticationResult authResult = context.acquireTokenByRefreshToken(
+                            refreshAuthResult,
+                            userInfo.getTenantId(),
+                            resource,
+                            settings.getClientId());
+                    validateAuthenticationResult(authResult);
+
+                    UserInfo userInfo2 = new UserInfo(authResult.getUserInfo().getTenantId(),
+                            authResult.getUserInfo().getUniqueName());
+
+                    if (!userInfo.equals(userInfo2)) {
+                        throw new AzureCmdException("Invalid User Information retrieved");
+                    }
+
+                    setAuthenticationResult(userInfo, resource, authResult);
+                } finally {
+                    if (context != null) {
+                        context.dispose();
+                    }
+                }
+            }
+        } catch (Throwable t) {
+            if (isErrorTokenUnauthorized(t)) {
+                authenticateWithInteractiveToken(userInfo, resource, title);
+            } else if (t instanceof AzureCmdException) {
+                throw (AzureCmdException) t;
+            } else if (t instanceof ExecutionException) {
+                throw new AzureCmdException("Error invoking the authentication process with refresh token", t.getCause());
+            } else {
+                throw new AzureCmdException("Error invoking the authentication process with refresh token", t);
+            }
+        } finally {
+            // we release the lock before we do anything else as
+            // we don't want a lock leaking in case the statements
+            // following throw
+            userLock.writeLock().unlock();
+        }
+    }
+
+    private void authenticateWithInteractiveToken(@NotNull UserInfo userInfo,
+                                                  @NotNull String resource,
+                                                  @NotNull String title)
+            throws AzureCmdException {
+        AuthenticationResult authenticationResult = isAuthenticated(userInfo, resource) ?
+                getAuthenticationResult(userInfo, resource) :
+                null;
+        AuthenticationResult refreshAuthResult = hasRefreshAuthResult(userInfo) ?
+                getRefreshAuthResult(userInfo) :
+                null;
+        boolean useRenewedRefresh = false;
+
+        ReentrantReadWriteLock userLock = getTempUserLock(userInfo);
```

Comment:
This is used for interactive requests for new UserInfo, where there might be no previous locks. If many happen concurrently, it will end up bringing many login windows; this is an unusual scenario, but it happened to me once.
It will create a UserLock only for interactive authentication.